sys/mips/pmap.c

```diff
@@ -307,6 +308,26 @@ static bool pmap_extract_nolock(pmap_t *pmap, vaddr_t va, paddr_t *pap) {
   return true;
 }
 
+bool pmap_extract_and_hold(pmap_t *pmap, vaddr_t va, paddr_t *pap,
+                           vm_prot_t prot) {
+  assert(pap != NULL);
+
+  SCOPED_MTX_LOCK(&pmap->mtx);
+
+  if (!pmap_extract_nolock(pmap, va, pap)) {
+    return false;
+  }
+
+  pte_t pte = pmap_pte_read(pmap, va);
+
+  if ((prot & VM_PROT_READ && !(pte & PTE_READ)) ||
+      (prot & VM_PROT_WRITE && !(pte & PTE_WRITE))) {
+    return false;
+  }
+
+  return true;
+}
```

Comment:
Do you really need to return pa? Maybe you can hide more details inside pmap module and don't expose them to other part of code. It should increase readability of `trap.c`.

Reply:
I think I need it. This physical address is read in `trap.c`, and then we look for the corresponding `vm_page_t`. Also returning pa allows to fix this bug only by changing as few as possible lines in `trap.c`.

Reply:
Why you can't hide software bits emulation into pmap module?

Reply:
I don't understand what do you mean exactly. Could you explain?

Reply:
Changed.